### PR TITLE
[WIP] Reformat and improve RAG module and agents

### PR DESCRIPTION
**Pull request by ZiTao-Li**
---
Reformat and improve RAG module and agents

---
## Description

* migrate and reformat RAG module(s) and RAG agent(s) from examples to a module in src
* support multiple loader

[WIP] add/remove document
[WIP] more easy-to-use RAG module config

## Checklist

Please check the following items before code is ready to be reviewed.

- [ ]  Code has passed all tests
- [ ]  Docstrings have been added/updated in Google Style
- [ ]  Documentation has been updated
- [ ]  Code is ready for review

### CODE REVIEW

**Comment by DavdGao** [comment]
@ZiTao-Li Is this PR ready for review?

File: docs/sphinx_doc/en/source/tutorial/209-rag.md [hunk 1]
```diff
@@ -0,0 +1,126 @@
+(209-rag-en)=
+
+# A Quick Introduction to RAG in AgentScope
+
+We want to introduce three concepts related to RAG in AgentScope: Knowledge, KnowledgeBank and RAG agent.
```
Comment (DavdGao) [inline docs/sphinx_doc/en/source/tutorial/209-rag.md:5]: If the document if for AgentScope developers, we should avoid to use the concepts in LlamaIndex (e.g. nodes, docs, etc), or provide pre knowledge here (if necessary).

File: docs/sphinx_doc/en/source/tutorial/209-rag.md [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+
+Here, we will use `LlamaIndexKnowledge` as an example to illustrate the operation within the `Knowledge` module.
+When a `LlamaIndexKnowledge` object is initialized, the `LlamaIndexKnowledge.__init__` will go through the following steps:
+  *  It processes documents and generates indexing for retrieval in `LlamaIndexKnowledge._data_to_index(...)`, which includes
+      * loading the documents `LlamaIndexKnowledge._data_to_docs(...)`;
+      * preprocessing the documents to nodes with preprocessing methods and embedding model `LlamaIndexKnowledge._docs_to_nodes(...)`;
+      * generating index with the processed nodes.
+  * If the indexing already exists, then `LlamaIndexKnowledge._load_index(...)` will be invoked to load the index and avoid repeating embedding calls.
+
+  A RAG module can be created with a JSON configuration to specify 1) data path, 2) data loader, 3) data preprocessing methods, and 4) embedding model (model config name).
```
Comment (DavdGao) [inline docs/sphinx_doc/en/source/tutorial/209-rag.md:18]: Is this RAG module the "Knowledge" module?

File: docs/sphinx_doc/en/source/tutorial/209-rag.md [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+  </details>
+
+If users want to avoid the detailed configuration, we also provide a quick way in `KnowledgeBank` (see the following).
+  </br>
+
+### Knowledge Bank
+The knowledge bank maintains a collection of Knowledge objects (e.g., on different datasets) as a set of *knowledge*. Thus,
+different agents can reuse the RAG modules without unnecessary "re-initialization".
+Considering that configuring the RAG module may be too complicated for most users, the knowledge bank also provides an easy function call to create RAG modules.
+  * `KnowledgeBank.add_data_as_knowledge`: create RAG module. An easy way only requires to provide `knowledge_id`, `emb_model_name` and `data_dirs_and_types`
```
Comment (DavdGao) [inline docs/sphinx_doc/en/source/tutorial/209-rag.md:83]: Similar as above. Does the "RAG module" refer to Knowledge class?

**Review by DavdGao — COMMENTED** [review]
It's hard to follow the tutorial to use the RAG module. Some important guidances are missing as follows. More details please refer to inline comments.
1. how to use the knowledge objects here, what methods does it provide? 
2. how to set the configuration, and what's the meaning and candidate values of each parameter?

File: examples/conversation_with_RAG_agents/rag_example.py [hunk 4]
```diff
@@ -1,63 +1,145 @@
...
+    guide_agent = agent_list[4]
+
+    # the knowledge bank can be configured by loading config file
+    with open(
+        "configs/knowledge_config.json",
+        "r",
+        encoding="utf-8",
+    ) as f:
+        knowledge_configs = json.load(f)
+    knowledge_bank = KnowledgeBank(configs=knowledge_configs)
```
Comment (qbc2016) [inline examples/conversation_with_RAG_agents/rag_example.py:84]: How about allowing the configuration to accept file paths as input?

File: src/agentscope/rag/knowledge_bank.py [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+                unique id for the RAG
+            duplicate (bool):
+                whether return a copy of the RAG.
+        Returns:
+            LlamaIndexRAG:
+                the RAG object defined with Llama-index
+        """
+        if knowledge_id not in self.stored_knowledge:
+            raise ValueError(
+                f"{knowledge_id} does not exist in the " f"knowledge bank.",
```
Comment (qbc2016) [inline src/agentscope/rag/knowledge_bank.py:145]: Please reformat this line.

File: src/agentscope/agents/rag_agents.py [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+                # to the user input.
+                msg = Msg(
+                    name="user",
+                    role="user",
+                    content=CHECKING_PROMPT.format(
+                        retrieved_docs_to_string,
+                        query,
+                    ),
+                )
+                print(msg)
```
Comment (qbc2016) [inline src/agentscope/agents/rag_agents.py:146]: Do we need `print(msg)` here?

File: src/agentscope/agents/rag_agents.py [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+from typing import Optional, Any
+from loguru import logger
+
+from agentscope.agents.agent import AgentBase
+from agentscope.message import Msg
+from agentscope.rag import Knowledge
+
+
+CHECKING_PROMPT = """
+                Does the retrieved content is relevant to the query?
```
Comment (qbc2016) [inline src/agentscope/agents/rag_agents.py:18]: Please correct the grammar issue.

File: docs/sphinx_doc/zh_CN/source/tutorial/209-rag.md [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+  knowledge_bank.add_data_as_knowledge(
+      knowledge_id=knowledge_config["knowledge_id"],
+      emb_model_name=knowledge_config["emb_model_config_name"],
+      knowledge_config=knowledge_config,
+  )
+  ```
+* `KnowledgeBank.get_knowledge`: 它接受两个参数，knowledge_id和duplicate。
+  如果duplicate为true，则返回提供的knowledge_id对应的知识对象；否则返回深拷贝的对象。
+* `KnowledgeBank.equip`: 它接受两个参数，`agent`和`duplicate`。
+该函数首先会检查代理是否具有rag_config；如果有，则根据rag_config中的knowledge_id提供相应的知识，并为代理初始化检索器。
```
Comment (qbc2016) [inline docs/sphinx_doc/zh_CN/source/tutorial/209-rag.md:104]: Perhaps we should standardize the translation of "agent" as "智能体"？

File: docs/sphinx_doc/zh_CN/source/tutorial/209-rag.md [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+`duplicate` 同样决定是否是深拷贝。
+
+
+### RAG agent
+RAG agent是可以基于检索到的知识生成答案的agent。
+  * 让Agent使用RAG: RAG agent在其配置中需要·`rag_config`，其中有一个`knowledge_id`的列表
```
Comment (qbc2016) [inline docs/sphinx_doc/zh_CN/source/tutorial/209-rag.md:110]: "·`rag_config`" has an extra dot.

File: examples/conversation_with_RAG_agents/rag_example.py [hunk 10]
```diff
@@ -1,63 +1,145 @@
...
+    with open("configs/agent_config.json", "r", encoding="utf-8") as f:
         agent_configs = json.load(f)
-    tutorial_agent = LlamaIndexAgent(**agent_configs[0]["args"])
-    code_explain_agent = LlamaIndexAgent(**agent_configs[1]["args"])
-    summarize_agent = DialogAgent(**agent_configs[2]["args"])
+
+    agent_list = agentscope.init(
+        model_configs=model_configs,
+        agent_configs=agent_configs,
+    )
```
Comment (qbc2016) [inline examples/conversation_with_RAG_agents/rag_example.py:73]: Shall we assume that the api keys have been set in the config? Thus, the code can be cleaner as 
```
agent_list = agentscope.init(
        model_configs="./configs/model_confi.json",
        agent_configs="./configs/agent_config.json",
    )
```

File: examples/conversation_with_RAG_agents/rag_example.py [hunk 11]
```diff
@@ -1,63 +1,161 @@
...
-            },
-        ],
+    # prepare html for api agent
+    prepare_docstring_html()
+
+    # prepare models
+    with open("configs/model_config.json", "r", encoding="utf-8") as f:
+        model_configs = json.load(f)
+    # for internal API
+    for config in model_configs:
```
[the remark this answers is not in the document]
Comment (ZiTao-Li) [reply examples/conversation_with_RAG_agents/rag_example.py]: removed

File: examples/conversation_with_RAG_agents/rag_example.py (hunk 10 above)
Comment (ZiTao-Li) [reply examples/conversation_with_RAG_agents/rag_example.py:64]: update as suggested

File: docs/sphinx_doc/zh_CN/source/tutorial/209-rag.md (hunk 9 above)
Comment (ZiTao-Li) [reply docs/sphinx_doc/zh_CN/source/tutorial/209-rag.md]: fixed

File: docs/sphinx_doc/zh_CN/source/tutorial/209-rag.md (hunk 8 above)
Comment (ZiTao-Li) [reply docs/sphinx_doc/zh_CN/source/tutorial/209-rag.md]: fixed

File: src/agentscope/agents/rag_agents.py (hunk 7 above)
Comment (ZiTao-Li) [reply src/agentscope/agents/rag_agents.py]: fixed

File: src/agentscope/agents/rag_agents.py (hunk 6 above)
Comment (ZiTao-Li) [reply src/agentscope/agents/rag_agents.py]: removed

File: src/agentscope/rag/knowledge_bank.py (hunk 5 above)
Comment (ZiTao-Li) [reply src/agentscope/rag/knowledge_bank.py]: done

File: examples/conversation_with_RAG_agents/rag_example.py (hunk 4 above)
Comment (ZiTao-Li) [reply examples/conversation_with_RAG_agents/rag_example.py]: update as suggested

File: docs/sphinx_doc/en/source/tutorial/209-rag.md (hunk 3 above)
Comment (ZiTao-Li) [reply docs/sphinx_doc/en/source/tutorial/209-rag.md]: fixed, removed the "RAG module" terms

File: docs/sphinx_doc/en/source/tutorial/209-rag.md (hunk 2 above)
Comment (ZiTao-Li) [reply docs/sphinx_doc/en/source/tutorial/209-rag.md]: fixed, removed the "RAG module" terms

File: docs/sphinx_doc/en/source/tutorial/209-rag.md (hunk 1 above)
Comment (ZiTao-Li) [reply docs/sphinx_doc/en/source/tutorial/209-rag.md:5]: Tried to update the terms/words to avoid requiring too much knowledge from LlamaIndex

File: src/agentscope/rag/llama_index_knowledge.py [hunk 12]
```diff
@@ -0,0 +1,567 @@
...
+        VectorStoreIndex,
+        StorageContext,
+        load_index_from_storage,
+    )
+    from llama_index.core.schema import (
+        Document,
+        TransformComponent,
+    )
+except ImportError:
+    BaseRetriever = None
```
Comment (garyzhang99) [inline src/agentscope/rag/llama_index_knowledge.py:31]: Should we consider referring to the approach in PR: https://github.com/modelscope/agentscope/pull/224 here? Use ImportErrorReporter to provide better import error message.

File: src/agentscope/rag/llama_index_knowledge.py [hunk 13]
```diff
@@ -0,0 +1,567 @@
...
+        self,
+        rag_config: Optional[dict] = None,
+        retriever: Optional[BaseRetriever] = None,
+        **kwargs: Any,
+    ) -> None:
+        """
+        Set the retriever as needed, or just use the default setting.
+
+        Args:
+            retriever (Optional[BaseRetriever]): passing a retriever in llama
```
Comment (garyzhang99) [inline src/agentscope/rag/llama_index_knowledge.py:429]: passing a retriever in llama-index

**Review by garyzhang99 — COMMENTED** [review]
Please see the inline comments.

File: src/agentscope/agents/rag_agents.py [hunk 14]
```diff
@@ -0,0 +1,181 @@
...
+                # to the user input.
+                msg = Msg(
+                    name="user",
+                    role="user",
+                    content=CHECKING_PROMPT.format(
+                        retrieved_docs_to_string,
+                        query,
+                    ),
+                )
+                checking = self.model([msg])
```
Comment (pan-x-c) [inline src/agentscope/agents/rag_agents.py:146]: should use `self.model.format` before calling model ?

File: src/agentscope/agents/rag_agents.py [hunk 15]
```diff
@@ -0,0 +1,181 @@
...
+    """
+
+    def __init__(
+        self,
+        name: str,
+        sys_prompt: str,
+        model_config_name: str,
+        knowledge_list: list[Knowledge] = None,
+        memory_config: Optional[dict] = None,
+        rag_config: Optional[dict] = None,
```
Comment (pan-x-c) [inline src/agentscope/agents/rag_agents.py:37]: declare a `RagConfig` class may be better

File: src/agentscope/rag/knowledge_bank.py [hunk 16]
```diff
@@ -0,0 +1,179 @@
...
+        """
+        Equip the agent with the knowledge it requests
+        Args:
+            agent (AgentBase): the agent to be equipped with knowledge
+                if it has "rag_config"
+            duplicate (bool): whether to deepcopy the knowledge object
+        """
+        if hasattr(agent, "rag_config") and "knowledge_id" in agent.rag_config:
+            if not hasattr(agent, "knowledge_list"):
+                agent.knowledge_list = []
```
Comment (pan-x-c) [inline src/agentscope/rag/knowledge_bank.py:169]: this method is not compatible with rpc agent.  RpcAgent only has reply and observe methods inherited from `AgentBase`, we need to add a function to `AgentBase` to equip agent with the knowledge.

**Review by pan-x-c — COMMENTED** [review]
Please see the inline comments.

The current version of rag is not compatible with distributed mode. We can add support in future PRs

File: docs/sphinx_doc/zh_CN/source/tutorial/209-rag.md [hunk 17]
```diff
@@ -0,0 +1,121 @@
+(209-rag-zh)=
+
+# 简要介绍AgentScope中的RAG
+
+我们在此介绍AgentScope与RAG相关的三个概念：知识（Knowledge），知识库（Knowledge Bank）和RAG 智能体。
+
+### Knowledge
+知识模块（目前仅有“LlamaIndexKnowledge”；即将支持对LangChain）负责处理所有与RAG相关的操作。
```
[the remark this answers is not in the document]
Comment (ZiTao-Li) [reply docs/sphinx_doc/zh_CN/source/tutorial/209-rag.md]: fixed

File: src/agentscope/rag/llama_index_knowledge.py (hunk 12 above)
Comment (ZiTao-Li) [reply src/agentscope/rag/llama_index_knowledge.py]: updated

File: src/agentscope/rag/llama_index_knowledge.py (hunk 13 above)
Comment (ZiTao-Li) [reply src/agentscope/rag/llama_index_knowledge.py]: fixed

File: src/agentscope/agents/rag_agents.py (hunk 14 above)
Comment (ZiTao-Li) [reply src/agentscope/agents/rag_agents.py]: added

File: src/agentscope/agents/rag_agents.py (hunk 15 above)
Comment (ZiTao-Li) [reply src/agentscope/agents/rag_agents.py]: done

File: src/agentscope/rag/knowledge_bank.py (hunk 16 above)
Comment (ZiTao-Li) [reply src/agentscope/rag/knowledge_bank.py:170]: added TODO to document RpcAgent support as future work

File: examples/conversation_with_RAG_agents/configs/agent_config.json [hunk 18]
```diff
@@ -0,0 +1,72 @@
...
+    }
+  },
+  {
+    "class": "LlamaIndexAgent",
+    "args": {
+      "name": "Code-Search-Assistant",
+      "description": "Code-Search-Assistant is an agent that can provide answer based on AgentScope code base. It can answer questions about specific modules in AgentScope.",
+      "sys_prompt": "You're a coding assistant of AgentScope. The answer starts with appreciation for the question, then provide details regarding the functionality and features of the modules mentioned in the question. The language should be in a professional and simple style. The answer is limited to be less than 100 words.",
+      "model_config_name": "qwen_config",
+      "rag_config": {
```
Comment (DavdGao) [inline examples/conversation_with_RAG_agents/configs/agent_config.json:24]: Can we do this pre-processing outside the agent? For example (taking `get` as example):
```python
knowledge_bank = KnowledgeBank(...)

knowledges = knowledge_bank.get(knowledge_ids=["kb1", "kb2"], similarity_top_k=5, log_retrieval=5, recent_n_mem=1)

AgentClass(name="assistant", knowledges=knowledges, ...)
```
or user can setup their own knowledge within the agent object's constructor by themselves. 

There are two advantages:
1. No need to know what parameters should be written in a rag config. All parameters are in the declaration of this `get()` function, which can be accessed easily.
2. The agent is not required to have a rag config attribute.

**Review by DavdGao — COMMENTED** [review]
please see inline comments

File: docs/sphinx_doc/zh_CN/source/tutorial/209-rag.md [hunk 19]
```diff
@@ -0,0 +1,121 @@
...
+
+  </details>
+
+如果用户想要避免详细的配置，我们也在`KnowledgeBank`中提供了一种快速的方式（请参阅以下内容）。
+  </br>
+
+### Knowledge Bank
+知识库将一组Knowledge模块（例如，来自不同数据集的知识）作为知识的集合进行维护。因此，不同的智能体可以在没有不必要的重新初始化的情况下重复使用知识模块。考虑到配置Knowledge模块可能对大多数用户来说过于复杂，知识库还提供了一个简单的函数调用来创建Knowledge模块。
+
+* `KnowledgeBank.add_data_as_knowledge`: 创建Knowledge模块。一种简单的方式只需要提供knowledge_id、emb_model_name和data_dirs_and_types。
```
[the remark this answers is not in the document]
Comment (ZiTao-Li) [reply docs/sphinx_doc/zh_CN/source/tutorial/209-rag.md:92]: added explanation.

File: examples/conversation_with_RAG_agents/README.md [hunk 20]
```diff
@@ -23,35 +22,28 @@ capability can be used to build easily.
 **Note:** This example has been tested with `dashscope_chat` and `dashscope_text_embedding` model wrapper, with `qwen-max` and `text-embedding-v2` models.
 However, you are welcome to replace the Dashscope language and embedding model wrappers or models with other models you like to test.
 
-## Start AgentScope Consultants
+## Start AgentScope Copilots
 * **Terminal:** The most simple way to execute the AgentScope Consultants is running in terminal.
```
[the remark this answers is not in the document]
Comment (ZiTao-Li) [reply examples/conversation_with_RAG_agents/README.md]: fixed

[the remark this answers is not in the document]
Comment (ZiTao-Li) [reply examples/conversation_with_RAG_agents/README.md]: fixed

File: docs/sphinx_doc/zh_CN/source/tutorial/209-rag.md [hunk 21]
```diff
@@ -0,0 +1,121 @@
...
+* `KnowledgeBank.get_knowledge`: 它接受两个参数，knowledge_id和duplicate。
+  如果duplicate为true，则返回提供的knowledge_id对应的知识对象；否则返回深拷贝的对象。
+* `KnowledgeBank.equip`: 它接受两个参数，`agent`和`duplicate`。
+该函数首先会检查智能体是否具有rag_config；如果有，则根据rag_config中的knowledge_id提供相应的知识，并为智能体初始化检索器。
+`duplicate` 同样决定是否是深拷贝。
+
+
+### RAG 智能体
+RAG 智能体是可以基于检索到的知识生成答案的智能体。
+  * 让智能体使用RAG: RAG agent在其配置中需要`rag_config`，其中有一个`knowledge_id`的列表
```
[the remark this answers is not in the document]
Comment (ZiTao-Li) [reply docs/sphinx_doc/zh_CN/source/tutorial/209-rag.md]: remove the `rag_config` and list all necessary parameter as agent level (as part of the agent's configuration)

File: docs/sphinx_doc/zh_CN/source/tutorial/209-rag.md [hunk 22]
```diff
@@ -0,0 +1,121 @@
...
+如果用户想要避免详细的配置，我们也在`KnowledgeBank`中提供了一种快速的方式（请参阅以下内容）。
+  </br>
+
+### Knowledge Bank
+知识库将一组Knowledge模块（例如，来自不同数据集的知识）作为知识的集合进行维护。因此，不同的智能体可以在没有不必要的重新初始化的情况下重复使用知识模块。考虑到配置Knowledge模块可能对大多数用户来说过于复杂，知识库还提供了一个简单的函数调用来创建Knowledge模块。
+
+* `KnowledgeBank.add_data_as_knowledge`: 创建Knowledge模块。一种简单的方式只需要提供knowledge_id、emb_model_name和data_dirs_and_types。
+  ```python
+  knowledge_bank.add_data_as_knowledge(
+        knowledge_id="agentscope_tutorial_rag",
```
[the remark this answers is not in the document]
Comment (FredericW) [reply docs/sphinx_doc/zh_CN/source/tutorial/209-rag.md:96]: ID (identity) could be either a name or number, as long as it is unique in the set-up.

File: examples/conversation_with_RAG_agents/configs/agent_config.json (hunk 18 above)
Comment (ZiTao-Li) [reply examples/conversation_with_RAG_agents/configs/agent_config.json]: Support after update. Now, the knowledge can be obtained by `get_knowledge` function, and a list of `knowledge` can be assigned to a agent in initialization.

In this update, agents are changed to use `knowledge.retrieve` function directly (the retriever is removed). The retriever is build in the `knowledge.retrieve` function every time called, with the parameter provided.

File: examples/conversation_with_RAG_agents/rag_example.py [hunk 23]
```diff
@@ -1,63 +1,146 @@
...
+            "".join(rag_agent_descriptions),
+        )
+    )
 
     user_agent = UserAgent()
-    # start the conversation between user and assistant
     while True:
+        # The workflow is the following:
+        # 1. user input a message,
+        # 2. if it mentions one of the agents, then the agent will be called
```
Comment (garyzhang99) [inline examples/conversation_with_RAG_agents/rag_example.py:123]: Should we tell the user that the word `mention` referred to the `@` operation here?

File: examples/conversation_with_RAG_agents/configs/agent_config.json [hunk 24]
```diff
@@ -0,0 +1,64 @@
...
+      "log_retrieval": false,
+      "recent_n_mem_for_retrieve": 1
+    }
+  },
+  {
+    "class": "LlamaIndexAgent",
+    "args": {
+      "name": "API-Assistant",
+      "description": "API-Assistant is an agent that can answer questions about APIs in AgentScope. It can answer general questions about AgentScope.",
+      "sys_prompt": "You're an assistant providing answers to the questions related to APIs (functions and classes) in AgentScope. The language style is helpful and cheerful. You generate answers based on the provided context. The answer is expected to be no longer than 200 words. If the key words of the question can be found in the provided context, the answer should contain the module of the API. For example, 'You may refer to MODULE_NAME for more details.'",
```
Comment (garyzhang99) [inline examples/conversation_with_RAG_agents/configs/agent_config.json:33]: This is not an issue, but I am curios that will the corresponding Chinese version of the Copilot prompt be included?

File: src/agentscope/rag/knowledge_bank.py [hunk 25]
```diff
@@ -0,0 +1,186 @@
...
+    KnowledgeBank enables
+    1) provide an easy and fast way to initialize the Knowledge object;
+    2) make Knowledge object reusable and sharable for multiple agents.
+    """
+
+    def __init__(
+        self,
+        configs: Union[dict, str],
+    ) -> None:
+        """initialize the knowledge bank"""
```
Comment (garyzhang99) [inline src/agentscope/rag/knowledge_bank.py:46]: Should we add some docs to show what fields are required in this config here? (e.g. ['knowledge_id"], ["emb_model_config_name"], etc.)

File: src/agentscope/rag/llama_index_knowledge.py [hunk 26]
```diff
@@ -0,0 +1,579 @@
...
+            }
+        logger.info("loaders are ready.")
+        return loader
+
+    def _set_transformations(self, config: dict) -> Any:
+        """
+        Set the transformations as needed, or just use the default setting.
+
+        Args:
+            config (dict): a dictionary containing configurations
```
Comment (garyzhang99) [inline src/agentscope/rag/llama_index_knowledge.py:395]: Not an issue, since the function won't expose to users, but it would be better if the `config` arg is more specified( e.g. the store_and_index field required?)

File: src/agentscope/agents/rag_agents.py [hunk 27]
```diff
@@ -0,0 +1,199 @@
...
+    """
+    A LlamaIndex agent build on LlamaIndex.
+    """
+
+    def __init__(
+        self,
+        name: str,
+        sys_prompt: str,
+        model_config_name: str,
+        memory_config: Optional[dict] = None,
```
Comment (DavdGao) [inline src/agentscope/agents/rag_agents.py:34]: Consider to remove memory_config since we never use it.

**Review by DavdGao — COMMENTED** [review]
plz see inline comments

File: src/agentscope/rag/llama_index_knowledge.py [hunk 28]
```diff
@@ -0,0 +1,579 @@
...
+    DEFAULT_CHUNK_OVERLAP,
+)
+from agentscope.rag.knowledge import Knowledge
+
+
+try:
+
+    class _EmbeddingModel(BaseEmbedding):
+        """
+        wrapper for ModelWrapperBase to an embedding model can be used
```
Comment (garyzhang99) [inline src/agentscope/rag/llama_index_knowledge.py:59]: TO DISCUSS: I actually think it would be useful, that providing  wrappers in `agentscope` to wrap the ModelWrapperBase so that it could be used in other pipelines such as `llama-Index` pipeline or `langchain` pipeline. Should we do it?

File: examples/conversation_with_RAG_agents/configs/agent_config.json (hunk 24 above)
Comment (FredericW) [reply examples/conversation_with_RAG_agents/configs/agent_config.json:33]: It will not be included here as this is just an example for demo, but we will add that to the Copilot project (close-sourced).

File: src/agentscope/rag/knowledge_bank.py (hunk 25 above)
Comment (FredericW) [reply src/agentscope/rag/knowledge_bank.py:46]: We have added remarks regarding such a format in the tutorial.

File: src/agentscope/rag/llama_index_knowledge.py (hunk 28 above)
Comment (FredericW) [reply src/agentscope/rag/llama_index_knowledge.py:59]: Good suggestion! As we discussed in the ding group, we will update this in the next release.

File: src/agentscope/agents/rag_agents.py (hunk 27 above)
Comment (FredericW) [reply src/agentscope/agents/rag_agents.py:34]: We may not remove it for the moment... We may need to use it for other applications in the future.

File: docs/sphinx_doc/en/source/tutorial/209-rag.md [hunk 29]
```diff
@@ -0,0 +1,158 @@
...
+### Knowledge Bank
+The knowledge bank maintains a collection of Knowledge objects (e.g., on different datasets) as a set of *knowledge*. Thus,
+different agents can reuse the Knowledge object without unnecessary "re-initialization".
+Considering that configuring the Knowledge object may be too complicated for most users, the knowledge bank also provides an easy function call to create Knowledge objects.
+  * `KnowledgeBank.add_data_as_knowledge`: create Knowledge object. An easy way only requires to provide `knowledge_id`, `emb_model_name` and `data_dirs_and_types`.
+    As knowledge bank process files as `LlamaIndexKnowledge` by default, all text file types are supported, such as `.txt`, `.html`, `.md`, `.csv`, `.pdf` and all code file like `.py`.  File types other than the text can refer to [LlamaIndex document](https://docs.llamaindex.ai/en/stable/module_guides/loading/simpledirectoryreader/).
+  ```python
+  knowledge_bank.add_data_as_knowledge(
+        knowledge_id="agentscope_tutorial_rag",
+        emb_model_name="qwen_emb_config",
```
[the remark this answers is not in the document]
Comment (FredericW) [reply docs/sphinx_doc/en/source/tutorial/209-rag.md:99]: It is emb_model_name

**Comment by DavdGao** [comment]
The ImportError of LlamaIndex library is still exposed to users who don't use RAG module.

**Review by DavdGao — APPROVED** [review]
LGTM
